slime/backends/megatron_utils/loss.py

```diff
@@ -403,27 +405,55 @@ def policy_loss_function(
 
     log_probs = log_probs_and_entropy["log_probs"]
 
-    if args.advantage_estimator == "gspo":
-        full_log_probs = [
-            all_gather_with_cp(log_prob, total_length, response_length)
-            for log_prob, total_length, response_length in zip(
-                log_probs, total_lengths, response_lengths, strict=False
-            )
-        ]
-        full_old_log_probs = [
-            all_gather_with_cp(old_log_prob, total_length, response_length)
-            for old_log_prob, total_length, response_length in zip(
-                old_log_probs, total_lengths, response_lengths, strict=False
-            )
-        ]
+    # Pre-gather log probs if needed by OPSM or GSPO to avoid duplicate gathering
+    cp_size = mpu.get_context_parallel_world_size()
+    need_full_log_probs = (args.enable_opsm or args.advantage_estimator == "gspo") and cp_size > 1
```

Comment:
```suggestion
    need_full_log_probs = (args.enable_opsm or args.advantage_estimator == "gspo")
```
it seems that the code below has deal with cp == 1.